src/flamenco/runtime/fd_blockstore.c

```diff
@@ -548,56 +621,206 @@ fd_blockstore_buffered_shreds_remove( fd_blockstore_t * blockstore, ulong slot )
+                                   ulong src_sz,
+                                   ulong write_off ) {
+
+  if ( FD_UNLIKELY( lseek( fd, (long)write_off, SEEK_SET ) == -1 ) ) {
+    FD_LOG_ERR(( "[%s] failed to seek to offset %lu", __func__, write_off ));
+  }
+  ulong wsz;
+  ulong remaining_sz = blockstore->fd_size_max - write_off;
+  if ( remaining_sz < src_sz ) {
+    int err = fd_io_write( fd, src, remaining_sz, remaining_sz, &wsz );
+    check_read_write_err( err );
+    write_off = FD_BLOCKSTORE_ARCHIVE_START;
+    if ( FD_UNLIKELY( lseek( fd, (long)write_off, SEEK_SET ) == -1 ) ) {
+      FD_LOG_ERR(( "[%s] failed to seek to offset %lu", __func__, write_off ));
+    }
+    err = fd_io_write( fd, src + remaining_sz, src_sz - remaining_sz, src_sz - remaining_sz, &wsz );
+    check_read_write_err( err );
+    write_off += wsz;
+  } else {
+    int err = fd_io_write( fd, src, src_sz, src_sz, &wsz );
+    check_read_write_err( err );
+    write_off += wsz;
+  }
+  if ( write_off >= blockstore->fd_size_max ) {
+    write_off = FD_BLOCKSTORE_ARCHIVE_START;
+  }
+  return write_off;
+}
+
+static void start_archive_write( fd_blockstore_t * blockstore, int fd, ulong lrw_cleared_end ) {
+  /* Invalidates the blocks that will be overwritten by marking them as free space */
+  if ( FD_UNLIKELY( lseek( fd, 0, SEEK_SET ) == -1 ) ) {
+    FD_LOG_ERR(( "[%s] failed to seek to start", __func__ ));
+  }
+  fd_blockstore_archive_metadata_t metadata = { .fd_size_max = blockstore->fd_size_max,
+                                                .mrw_end = blockstore->off,
+                                                .lrw_off = lrw_cleared_end };
+  ulong wsz;
+  int err = fd_io_write( fd, &metadata, sizeof(ulong), sizeof(ulong), &wsz );
+  check_read_write_err( err );
+}
+
+static void end_archive_write( fd_blockstore_t * blockstore,
+                               int fd ) {
+  if ( FD_UNLIKELY( lseek( fd, 0, SEEK_SET ) == -1 ) ) {
+    FD_LOG_ERR(( "[%s] failed to seek to start", __func__ ));
+  }
+  fd_block_idx_t * block_idx = fd_blockstore_block_idx( blockstore );
+  fd_block_idx_t * lrw_block_index = fd_block_idx_query( block_idx, blockstore->lrw_slot, NULL);
+  fd_blockstore_archive_metadata_t metadata = { .fd_size_max = blockstore->fd_size_max,
+                                                .mrw_end = blockstore->off, 
+                                                .lrw_off = lrw_block_index->off };
+  ulong wsz;
+  int err = fd_io_write( fd, &metadata, sizeof(metadata), sizeof(metadata), &wsz );
+  check_read_write_err( err );
+}
+
+/* Clears any to be overwritten blocks in the archive from the index and updates lrw slot */
+static ulong
+fd_blockstore_lrw_archive_clear( fd_blockstore_t * blockstore, ulong slot, ulong wsz, ulong write_off ) {
+  fd_block_idx_t * block_idx = fd_blockstore_block_idx( blockstore );
+  fd_block_idx_t * lrw_block_index = fd_block_idx_query(block_idx, blockstore->lrw_slot, NULL);
```

Comment:
what if lrw_slot is null?

Reply:
then lrw_block_index would be null, which is handled -- but I'll change the check to be on the slot itself so its more clear